proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,465 @@
+~~~~~~~~~~~~~~~~
+
+Kind signatures for regular data type declarations must have a return kind of
+``TYPE (BoxedRep Lifted)``. Activating ``-XUnliftedDatatypes`` will lift this
+restriction to allow any return kind that unifies with ``TYPE (BoxedRep _)``. A
+data type constructor explitly *may* have levity polymorphic kind
+``forall l. TYPE (BoxedRep l)``.
+
+Note that Haskell98-style data declarations can use top-level kind signatures
+to specify the return kind. If the user provides no kind signature, the default
+remains that the data type has lifted return kind (``TYPE (BoxedRep Lifted)``).
+Specifying a kind signature is the only way to declare an unlifted data type.
+
+The same applies to data family instances, in which case the data family
+application's result kind must reduce to ``TYPE (BoxedRep _)``. See
+`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
+and ``Note [Implementation of UnliftedNewtypes]`` for details involving
+type-checking the parent data family.
+
+As usual for types of unlifted kind, they may not be declared at the top-level.
+The top-level restriction is not fundamental (see
+`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_), but best
+discussed in a separate proposal. There is an interesting interaction of this
+restriction with nullary data-con wrappers (cf. "Effects and Interactions"),
+which we consider to be an implementation detail.
+
+Dynamic Semantics
+~~~~~~~~~~~~~~~~~
+
+Unliftedness (i.e., the absence of divergence) in general implies the need for
+an eager evaluation semantics, which GHC implements in expressions of kind
+``TYPE r``, where ``r`` is not ``BoxedRep Lifted`` (so
+``TYPE (BoxedRep Unlifted)`` in particular).
+
+Thus, call-by-value semantics are already well established within GHC. The
+novelty is pattern matching on and construction of unlifted data types, but
+that's exactly the same as it is for lifted data types. 
+
+We never have to decide on an evaluation strategy for ``TYPE (BoxedRep l)``
+with inconcrete ``l``, because the current restrictions on levity polymorphism
+don't allow to bind or pass expressions of this kind.
+
+Examples
+--------
+
+Unlifted (call-by-value) semantics
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Example:
+
+::
+
+ data UPair a b :: TYPE (BoxedRep Unlifted) where
+   UPair :: a -> b -> UPair a b
+
+* When occuring in a constructor field (e.g.
+  ``data T = MkT (UPair Int Bool)``), the semantics are identical to a field
+  with a bang pattern (``data T = MkT !(Int, Bool)``).
+
+* In an application ``f (UPair a b)``, the argument ``UPair a b`` is evaluated
+  before the application is beta reduced. So call-by-value instead of
+  call-by-need.
+
+* In a let binding ``let x = UPair a b in e``, the right-hand side ``UPair a b``
+  is evaluated before the body.
+
+The Strict data type
+~~~~~~~~~~~~~~~~~~~~
+
+We get to define ``Strict``
```

Comment:
Why does this have to be a `data` type, allocating an extra box? Why can't it be a newtype? I know that newtypes typically have the same kind as their argument, but it actually seems just fine to have a newtype of kind `BoxedRep Unlifted` wrap a representation of kind `BoxedRep Lifted`. The type system would make sure that the value of type `StrictNewtype` would be evaluated eagerly. Perhaps that's a separate proposal, but one worth considering next, I think.